text/0000-get-expr-type.md

```diff
@@ -0,0 +1,239 @@
+- Feature Name: `get_expr_type`
+- Start Date: 2019-05-30
+- RFC PR: [rust-lang/rfcs#2706](https://github.com/rust-lang/rfcs/pull/2706)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add the ability to retrieve the concrete type of an arbitrary expression so that
+it may be reused by other code that relies on the expression.
+
+# Motivation
+[motivation]: #motivation
+
+> "Macros will become more powerful than you can possibly imagine."
+>
+> – Obi-Wan Kenobi (_supposedly_)
+
+Within the context of a macro, this feature would be very useful.
+
+It would allow for:
+
+- Defining a function that takes the concrete type of a given expression.
+- Calling upon type-level functions based upon a given expression.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given some value `x`, we can retrieve its type with just `x.type`. This type can
+then be used within various contexts:
+
+```rust
+let x = "hellooo";
+let y: x.type = x; // C equivalent: typeof(x) y = x;
+
+type X = x.type;
+
+assert_eq_type!(X, &str); // taken from `static_assertions`
+
+assert!(<x.type>::default().is_empty());
+```
+
+Note that if the expression resolves to a long operation, the operation will
+_not_ be evaluated.
```

Comment:
How does this interact with trait objects where the concrete type is not known at compile time?

Reply:
`dyn Trait` is the type.  Also `dyn Trait` reveals only what `Trait` permits about original type, although some traits like `Any` or `Error` provide `downcast` or `type_id` methods.

Reply:
is `dyn Trait` considered a concrete type? I would have expected the underlying type to be. Either it returns `dyn Trait` or is disallowed if the underlying type cannot be known at compile time - either way I think it deserves mention in the RFC.

Reply:
I've no idea if concrete means anything.  `dyn Trait` is an unsized type with a representation expressed in terms of pointer metadata, so vaguely like `[T]`.